pkg/storage/errors.go

```diff
@@ -18,8 +18,46 @@ package storage
 
 import (
 	etcdutil "k8s.io/kubernetes/pkg/storage/etcd/util"
+	"k8s.io/kubernetes/pkg/util/validation/field"
 )
 
+type InvalidError struct {
+	Errs field.ErrorList
+}
+
+func (e InvalidError) Error() string {
+	return e.Errs.ToAggregate().Error()
+}
+
+// IsInValidError returns true if and only if err is an InternalError.
+func IsInvalidError(err error) bool {
+	_, ok := err.(InvalidError)
+	return ok
+}
+
+func NewInvalidError(errors field.ErrorList) InvalidError {
+	return InvalidError{errors}
+}
+
+// InternalError represents errors in storage package that is not returned by etcd.
```

Comment:
"InternalError is generated when an error occurs in the storage package, i.e., not from the underlying storage backend (e.g., etcd)."